include/websocket.hpp

```diff
@@ -0,0 +1,472 @@
+
+    // Set a decorator to change the User-Agent of the handshake
+    ws_.set_option(
+        websocket::stream_base::decorator([](websocket::request_type &req) {
+          req.set(http::field::user_agent,
+                  std::string(BOOST_BEAST_VERSION_STRING) +
+                      " websocket-client-async");
+        }));
+
+    // Perform the websocket handshake
+    ws_.async_handshake(
+        host_, "/",
+        beast::bind_front_handler(&session::on_handshake, shared_from_this()));
+  }
+
+  /// @brief Send the ping message once the handshake has been complete calls ->
+  /// on_read
+  /// @param ec error code on handshake
+  void on_handshake(beast::error_code ec) {
+    if (ec) return fail(ec, "handshake");
+
+    // If you reach here connection is up set is_connected to true
+    is_connected = true;
+
+    // Start listening to the socket for messages
+    ws_.async_read(buffer_, beast::bind_front_handler(&session::on_read,
+                                                      shared_from_this()));
+  }
+
+  /// @brief Deals with the message recieved from server, calls itself to keep
+  /// listening to more messages
+  /// @param ec error code while reading
+  /// @param bytes_transferred Amount of byte recieved
+  void on_read(beast::error_code ec, std::size_t bytes_transferred) {
+    boost::ignore_unused(bytes_transferred);
+    // report error on read
+    if (ec) {
+      return fail(ec, "read");
+    }
+
+    // id socket has been disconnected then return
+    if (!is_connected) {
+      return;
+    }
+
+    // get the data from the websocket and parse it to json
+    auto res = buffer_.data();
+    json data = json::parse(net::buffers_begin(res), net::buffers_end(res));
+    // std::cout << data << std::endl;
+    // denotes that this response is just after writing for subscription
+    if (justSubscribed) {
+      add_callback(data);
+    }
+
+    // denotes that this response is just after writing for unsubscription
+    else if (justUnsubscribed) {
+      remove_callback(data);
+    }
+
+    // this is a notification process it sccordingly
+    else {
+      call_callback(data);
+    }
+
+    // consume the buffer
+    buffer_.consume(bytes_transferred);
+
+    // call read again to read continiously
+    ws_.async_read(buffer_, beast::bind_front_handler(&session::on_read,
+                                                      shared_from_this()));
+  }
+
+  /// @brief function to write to the websocket
+  void do_write() {
+    // if pendingSub and pendingUnSub is empty that meaans nothing to write stop
+    // thread execution if justSubscribed or justUnSubscribed i.e there is one
+    // async_write don't write again
+    while (is_connected and
+           ((pendingSub.size() == 0 and pendingUnSub.size() == 0) or
+            justSubscribed or justUnsubscribed))
+      ;
+
+    // if not connected then wait for 5s for connection
+    // if still not connected return
+    if (!is_connected) {
```

Comment:
Not a good way to wait for 5s, with this thread will be active all the time instead of going to sleep.